ui/vty/src/Ariadne/UI/Vty/Widget.hs

```diff
@@ -0,0 +1,464 @@
+
+-- | Widget rendering monad
+--
+-- Reader monad is used, because Brick doesn't allow changing app state during rendering.
+-- Local widget state @s@ is passed as a parameter to drawing routine
+-- to allow more convenient utilizing of @RecordWildcards@.
+type WidgetDrawM s p a = ReaderT (WidgetInfo s p) (Reader p) a
+
+-- | Widget event handler monad
+--
+-- Allows access to:
+-- * @s@: widget-local state, lenses just work
+-- * @WidgetInfo s p@: widget itself
+-- * @p@: parent widget via lenses and @widgetEvent@
+-- * @B.EventM@: Brick event handling environment
+type WidgetEventM s p a = StateT s (StateT (WidgetInfo s p) (StateT p (B.EventM WidgetName))) a
+
+makeLensesWith postfixLFields ''WidgetInfo
+
+----------------------------------------------------------------------------
+-- Widget initialization
+----------------------------------------------------------------------------
+
+-- | Creates an empty widget with default implementations of all routines
+-- and runs supplied constructor on it, returning opaque initialized widget
+initWidget :: WidgetInitM s p -> Widget p
+initWidget action =
+  Widget $ execState action WidgetInfo
+    { widgetName = []
+    , widgetState = error "State not defined for this widget"
+    , widgetChildren = Map.empty
+    , widgetFocusList = []
+    , widgetEventHandlers = Map.empty
+    , widgetEventQueue = []
+    , widgetEventSend = return . const ()
+    , widgetDraw = return . const B.emptyWidget
+    , widgetDrawWithFocused = \_ s -> do
+        draw <- view widgetDrawL
+        draw s
+    , widgetDrawWithFocus = \focus s -> do
+        draw <- view widgetDrawWithFocusedL
+        name <- view widgetNameL
+        draw (name == focus) s
+    , widgetHandleEditKey = return . const WidgetEventNotHandled
+    , widgetHandleKey = return . const WidgetEventNotHandled
+    , widgetHandlePaste = return . const WidgetEventNotHandled
+    , widgetHandleMouseDown = return . const WidgetEventNotHandled
+    , widgetHandleScroll = return . const WidgetEventNotHandled
+    , widgetHandleEvent = return . const ()
+    }
+
+setWidgetState :: s -> WidgetInitM s p
+setWidgetState = assign widgetStateL
+
+-- | Adds an opaque child to current widget
+--
+-- As child is created before parent (which will probably be changed at some point),
+-- the child has to be renamed to include path to parent.
+addWidgetChild :: WidgetNamePart -> Widget (WidgetInfo s p) -> WidgetInitM s p
+addWidgetChild namePart (Widget child) = do
+    let child' = child{ widgetEventSend = \event -> lift . lift $ widgetEventQueueL %= ((namePart, event):) }
+    widgetChildrenL %= Map.insert namePart (rename $ Widget child')
+  where
+    rename :: Widget p -> Widget p
+    rename (Widget widget@WidgetInfo{..}) =
+      Widget widget
+        { widgetName = namePart : widgetName
+        , widgetChildren = Map.map rename widgetChildren
+        }
+
+setWidgetFocusList :: MonadState (WidgetInfo s p) m => [WidgetNamePart] -> m ()
```

Comment:
When we get back to the topic of nested `StateT`s, I'd like to also consider using concrete monads here instead of `MonadState` constraints.